File: algorithms.rb
# fibonacci iterative
def fib_it(num)
  array = [0, 1]
  count = 2
  until count > num
    array << array[-2] + array[-1]
    count += 1
  end
  array[num]
end
# fib_it(5) output: 5

# fibonacci recursion
def fib(num)
  return num if num == 0 || num == 1
  fib(num - 1) + fib(num - 2)
end
# fib(5) output: 5

# simple string reverse without .reverse
def hard_reverse(str)
  str.chars.reduce{ |x,y| y+x }
end
# hard_reverse("hello") output: "olleh"

# palindrome check
def palindrome?(str)
  str == str.reverse
end
# palindrome?("wow") output: true
# palindrome?("hello") output: false

# binary search recursion with 4 args, returns index of obj
def binary_search(array, obj, min = 0, max = arr.length-1)
  mid = array[min..max].length / 2 + min
  return nil if min == max
  return mid if array[mid] == obj

  if array[mid] < obj
    binary_search(array, obj, mid, max)
  elsif array[mid] > obj
    binary_search(array, obj, min, mid)
  end
end
# array = [:A, :B, :C, :D, :E, :F, :G, :H]
# binary_search(array, :E)


# fizzbuzz with 15
def fizzbuzz?(n)
  case
  when n % 15 == 0 then "FizzBuzz"
  when n % 5 == 0 then "Buzz"
  when n % 3 == 0 then "Fizz"
  end
end

# fizzbuzz without 15
def fizzbuzz?(n)
  case
  when n % 5 && n % 3 == 0 then "FizzBuzz"
  when n % 5 == 0 then "Buzz"
  when n % 3 == 0 then "Fizz"
  end
end


# bubble sort
def bubble_sort(arr)
  n = arr.length
  loop do
    swapped = false

    (n-1).times do |i|
      if arr[i] > arr[i+1]
        arr[i], arr[i+1] = arr[i+1], arr[i]
        swapped = true
      end
    end

    break if not swapped
  end

  array
end


# quicksort
class Array
  def quicksort
    return [] if empty?

    pivot = delete_at(rand(size))
    left, right = partition(&pivot.method(:>))

    return *left.quicksort, pivot, *right.quicksort
  end
end
# arr = [34, 2, 1, 5,3]
# p arr.quicksort


# merge sort
def merge_sort(array)
  return array if array.size <= 1

  left, right = array.each_slice((array.size/2).round).to_a

  merge = -> (left, right) {
    array = []
    while left.size > 0 && right.size > 0
      array << (left[0] < right[0] ? left.shift : right.shift)
    end

    return array = array + left + right
  }

  return merge.call merge_sort(left), merge_sort(right)
end


# mirror symmetry
def is_symmetric(root)
  return true if root.nil?
  mirror(root.left, root.right)
end

def mirror(a,b)
  return a == b if (a == nil || b == nil)
  a.val == b.val && mirror(a.left, b.right) && mirror(a.right,b.left)
end

File: algorithms.js
// linear search
var linearSearch = function(array, obj) {
  for (var i = 0; i < array.length; ++i) {
    if (array[i] == obj) {
      return i;
    }
  }
  return -1;
}
// var array = [1,2,3,4,5,6,7,8,9,10];
// var obj = 6;
// linearSearch(array, obj) output: returns index 5 for obj


// binary search recursive
var binarySearch = function(array, obj, min, max) {
  var mid = Math.floor((max-min)/2);

  if (array[mid] === obj) {
    return mid;
  } else if (max === min) {
    return null;
  } else if (array[mid] < obj) {
    return binarySearch(array, obj, mid, max);
  } else if (array[mid] > obj) {
    return binarySearch(array, obj, min, mid)
  }
}
// var array = [1,2,3,4,5,6,7,8,9,10];
// var obj = 3;
// var min = 0;
// var max = array.length-1
// binarySearch(array, obj, min, max) output: returns index 2 for obj


// palindrome check
var palindrome = function(string) {
  return string == string.split('').reverse().join('');
}
// palindrome("hello") output: false
// palindrome("wow") output: true


// bubble sort
var bubbleSort = function(array) {
  for (var i = array.length - 1; i >= 0; i--) {
    for (var j = 1; j <= i; j++) {
      if (array[j-1] > array[j]) {
        var temp = array[j-1];
        array[j-1] = array[j];
        array[j] = temp;
      }
    }
  }
  return array
}
// bubbleSort([6,8,1,2,4,5]) output: [ 1, 2, 4, 5, 6, 8 ]


// merge sort in 2 functions
var mergeSort = function(list) {
  if (list.length <= 1)
    return list;

  var mid   = Math.floor(list.length / 2);
  var left  = list.slice(0, mid);
  var right = list.slice(mid, list.length);

    return merge(mergeSort(left), mergeSort(right))
}

var merge = function(left, right) {
  var temp = [];
  while (left && left.length > 0 && right && right.length > 0) {
    temp.push(left[0] <= right[0]? left.shift() : right.shift());
  }
  return temp.concat(left, right);
}
// nums = [10,9,8,7,6,5,5,5,4,3,2,1]
// mergeSort(nums) output: [ 1, 2, 3, 4, 5, 5, 5, 6, 7, 8, 9, 10 ]
